### PR TITLE
bank: username and password validation. filter with redirect.

### PR DESCRIPTION
**Pull request by martinmilev**
Added security filter. Added server side validation for username and password.

### CODE REVIEW

File: src/main/java/com/clouway/core/SessionsRepository.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+
+import java.util.Optional;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface SessionsRepository {
+  void save(String id);
+  Optional<Session> get();
```
Comment (mgenov) [inline src/main/java/com/clouway/core/SessionsRepository.java]: You don't need this interface and implementation in this PR cause the sessions.get() method is not used in the production code. So this piece could be completely removed and just random SessionId to be assigned as cookie. 

A better design of this interface could be: 

```java
public interface Sessions  {
  Session issueNewSession(String userId, Date instant); 
  Optional<Session> findSesssionAvailableAt(String userId, Date instant); 
}
```

The time is passed to both methods to make testing easy. 

You can keep it just as note for following PR.

File: src/main/java/com/clouway/http/filters/SecurityFilter.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    if(sessionIsNew(req)) {
+      resp.sendRedirect("login");
+    }
+
+  }
+
+  private boolean sessionIsNew(HttpServletRequest req) {
+    Cookie[] cookies = req.getCookies();
+    for (Cookie c : cookies) {
+      if (c.getName().equals("sid")) {
```
Comment (mgenov) [inline src/main/java/com/clouway/http/filters/SecurityFilter.java]: Most apps are using upper case characters for the session, so "SID" looks much common.

File: src/main/java/com/clouway/http/servlets/LoginPageServlet.java [hunk 3]
```diff
@@ -45,20 +61,40 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+      sessions.save(id);
+      Cookie cookie = new Cookie("sid", id);
+      resp.addCookie(cookie);
       resp.sendRedirect("/account");
     } else {
       servletResponseWriter.renderPage("login.html", Collections.singletonMap("error", "Wrong password"), resp);
     }
   }
 
+  static String sha1(String input) {
```
Comment (mgenov) [inline src/main/java/com/clouway/http/servlets/LoginPageServlet.java]: You can replace sha1 sessionId generation with simplified Base64 encoding + UUID:

```java
    UUID uuid = UUID.randomUUID();
    String newSesssionId = BaseEncoding.base64Url().encode(
            ByteBuffer.allocate(16).putLong(uuid.getMostSignificantBits()).putLong(uuid.getLeastSignificantBits()).array()
    );
```

which looks much simpler then it is.

File: src/main/java/com/clouway/http/servlets/LoginPageServlet.java [hunk 4]
```diff
@@ -45,20 +61,40 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+    if (!possibleAccount.isPresent() || !validator.check(name, "[a-zA-Z]{1,50}")) {
       servletResponseWriter.renderPage("login.html", Collections.singletonMap("error", "Wrong username"), resp);
       return;
     }
 
     Account account = possibleAccount.get();
 
-    if (pswd.equals(account.password)) {
+    if (pswd.equals(account.password) || !validator.check(pswd, "[a-zA-Z_0-9]{6,18}")) {
+      String id = sha1(UUID.randomUUID().toString());
```
Comment (mgenov) [inline src/main/java/com/clouway/http/servlets/LoginPageServlet.java]: Changing method name to:
```java
String newSessionId = newSessionId();
```
hides real algorithm by using common name which better describes the intention.

File: src/main/java/com/clouway/http/filters/SecurityFilter.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  public void init(FilterConfig filterConfig) throws ServletException {
+  }
+
+  @Override
+  public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
+
+    HttpServletRequest req = (HttpServletRequest) servletRequest;
+    HttpServletResponse resp = ((HttpServletResponse) servletResponse);
+
+    if(sessionIsNew(req)) {
```
Comment (mgenov) [inline src/main/java/com/clouway/http/filters/SecurityFilter.java]: sessionIsNew doesn't sound good. Change it to something like: 
```java
if (!isAuthorized(req)) {
  resp.sendRedirect("/login");
}
```

File: src/main/java/com/clouway/core/SessionsRepository.java (hunk 1 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsRepository.java]: fixed

File: src/main/java/com/clouway/http/filters/SecurityFilter.java (hunk 5 above)
Comment (martinmilev) [inline src/main/java/com/clouway/http/filters/SecurityFilter.java]: fixed

File: src/main/java/com/clouway/http/filters/SecurityFilter.java (hunk 2 above)
Comment (martinmilev) [inline src/main/java/com/clouway/http/filters/SecurityFilter.java]: fixed

File: src/main/java/com/clouway/http/servlets/LoginPageServlet.java (hunk 3 above)
Comment (martinmilev) [inline src/main/java/com/clouway/http/servlets/LoginPageServlet.java]: fixed

File: src/main/java/com/clouway/core/SessionsRepository.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.clouway.core;
+
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface SessionsRepository {
```
Comment (mgenov) [inline src/main/java/com/clouway/core/SessionsRepository.java]: Check my comment again !

File: src/main/java/com/clouway/core/Validator.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.clouway.core;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface Validator<T> {
```
Comment (mgenov) [inline src/main/java/com/clouway/core/Validator.java]: I've looked into the implementation of this interface and it's strange to me that value and matcher are of the same type. 

Also by passing matcher to the caller you are letting him know how data is going to be validated and this is not looking like a delegation for validation which is and the role of the validator.

```java
public interface Validator {
  // string as all param are coming in string format from the request
   boolean check(String value); 
}

class RegexValidator implements Validator<String> {
   public RegexValidator(String matcher) {
      ..
  }
  public boolean check(String value) {
      return matcher.matches(value); 
  }
}
...
RegexValidator userNameValidator = new RegexValidator("[a-zA-Z]{1,50}");
```

File: schemas/Bank_dll.sql [hunk 8]
```diff
@@ -3,4 +3,4 @@ ID        INT            NOT NULL PRIMARY KEY AUTO_INCREMENT,
 Name      VARCHAR(50)    NOT NULL,
 Password  VARCHAR(18)    NOT NULL,
 Amount    DOUBLE         NOT NULL
-);
```
Comment (mgenov) [inline schemas/Bank_dll.sql]: This change should be reverted as it's duplicated with #106

File: src/main/java/com/clouway/core/SessionsRepository.java (hunk 1 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsRepository.java]:
```java
Session issueNewSession(String userId, Date instant); 
```
Why do I need to return Session?
I was going to use that method to save the session to the DB.

**Comment by mgenov** [comment]
CI fails with compilation issue which need to be resolved:
https://travis-ci.org/clouway-incubator/servlets/builds/174499595

> [ERROR] /home/travis/build/clouway-incubator/servlets/src/test/java/com/clouway/http/servlets/LoginPageServletTest.java:[8,24] cannot find symbol
  symbol:   class SessionsRepository
  location: package com.clouway.core
[ERROR] /home/travis/build/clouway-incubator/servlets/src/test/java/com/clouway/http/servlets/LoginPageServletTest.java:[34,11] cannot find symbol
  symbol:   class SessionsRepository
  location: class com.clouway.http.servlets.LoginPageServletTest
[ERROR] /home/travis/build/clouway-incubator/servlets/src/test/java/com/clouway/http/servlets/LoginPageServletTest.java:[34,54] cannot find symbol
  symbol:   class SessionsRepository
  location: class com.clouway.http.servlets.LoginPageServletTest

File: src/main/webapp/WEB-INF/web.xml [hunk 9]
```diff
@@ -6,6 +6,20 @@
 
 <web-app>
 
+  <filter>
+    <filter-name>SecurityFilter</filter-name>
+    <filter-class>com.clouway.http.filters.SecurityFilter</filter-class>
+    <init-param>
```
Comment (mgenov) [inline src/main/webapp/WEB-INF/web.xml]: This param is not used and it need to be removed I suppose.

**Comment by mgenov** [comment]
You need to rebase your branch with upstream.

**Pull request merged**